src/test/socket/sendto_recvfrom/test_sendto_recvfrom.rs

```diff
@@ -0,0 +1,1379 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+use test_utils::TestEnvironment as TestEnv;
+use test_utils::{set, AsMutPtr, AsPtr};
+
+struct SendtoArguments<'a> {
+    fd: libc::c_int,
+    buf: Option<&'a [u8]>,
+    len: libc::size_t,
+    flags: libc::c_int,
+    addr: Option<libc::sockaddr_in>,
+    addr_len: libc::socklen_t,
+}
+
+struct RecvfromArguments<'a> {
+    fd: libc::c_int,
+    buf: Option<&'a mut [u8]>,
+    len: libc::size_t,
+    flags: libc::c_int,
+    addr: Option<libc::sockaddr_in>,
+    addr_len: Option<libc::socklen_t>,
+}
+
+impl Default for SendtoArguments<'_> {
+    fn default() -> Self {
+        Self {
+            fd: -1,
+            buf: None,
+            len: 0,
+            flags: 0,
+            addr: None,
+            addr_len: 0,
+        }
+    }
+}
+
+impl Default for RecvfromArguments<'_> {
+    fn default() -> Self {
+        Self {
+            fd: -1,
+            buf: None,
+            len: 0,
+            flags: 0,
+            addr: None,
+            addr_len: None,
+        }
+    }
+}
+
+fn main() -> Result<(), String> {
+    // should we restrict the tests we run?
+    let filter_shadow_passing = std::env::args().any(|x| x == "--shadow-passing");
+    let filter_libc_passing = std::env::args().any(|x| x == "--libc-passing");
+    // should we summarize the results rather than exit on a failed test
+    let summarize = std::env::args().any(|x| x == "--summarize");
+
+    let mut tests = get_tests();
+    if filter_shadow_passing {
+        tests = tests
+            .into_iter()
+            .filter(|x| x.passing(TestEnv::Shadow))
+            .collect()
+    }
+    if filter_libc_passing {
+        tests = tests
+            .into_iter()
+            .filter(|x| x.passing(TestEnv::Libc))
+            .collect()
+    }
+
+    test_utils::run_tests(&tests, summarize)?;
+
+    println!("Success.");
+    Ok(())
+}
+
+fn get_tests() -> Vec<test_utils::ShadowTest<(), String>> {
+    let mut tests: Vec<test_utils::ShadowTest<_, _>> = vec![];
+
+    tests.extend(vec![
```

Comment:
Nit - why not use this `vec!` above to construct the original `Vec`, rather than constructing an empty one and then extending it?

Reply:
Done. Not sure why I wrote it this way.